### ISSUE TITLE
Tests fail in xSPUserProfileSyncConnection under latest versions of Pester

### ISSUE THREAD

**Issue by BrianFarnhill** [PowerShell/xSharePoint#207]
Under Pester 3.4 or higher we are seeing failures in tests for xSPUserProfileSyncConnection. And example of the test output is below.

    Describing xSPUserProfileSyncConnection 
       Context When connection doesn't exist 
        [+] returns null from the Get method 302ms 
        [+] returns false when the Test method is called 47ms 
        [-] creates a new service application in the set method 148ms 
          RuntimeException: You cannot call a method on a null-valued expression. 
          at <ScriptBlock>, C:\projects\xsharepoint\Modules\xSharePoint\DSCResources\MSFT_xSPUserProfileSyncConnection\MSFT_xSPUserProfileSyncConnection.psm1: line 154 
          at <ScriptBlock>, <No file>: line 2 
          at <ScriptBlock>, C:\Program Files\WindowsPowerShell\Modules\Pester\Functions\Mock.ps1: line 1056 
          at ExecuteBlock, C:\Program Files\WindowsPowerShell\Modules\Pester\Functions\Mock.ps1: line 1060 
          at Invoke-Mock, C:\Program Files\WindowsPowerShell\Modules\Pester\Functions\Mock.ps1: line 906 
          at <ScriptBlock><Process>, <No file>: line 70 
          at Set-TargetResource, C:\projects\xsharepoint\Modules\xSharePoint\DSCResources\MSFT_xSPUserProfileSyncConnection\MSFT_xSPUserProfileSyncConnection.psm1: line 85 
          at <ScriptBlock>, C:\projects\xsharepoint\Tests\xSharePoint\xSharePoint.xSPUserProfileSyncConnection.Tests.ps1: line 145 
       Context When connection exists and account is different 
        [+] returns service instance from the Get method 155ms 
        [+] returns false when the Test method is called 93ms 
        [+] execute update credentials 44ms 
       Context When connection exists and forest is different 
        [+] returns service instance from the Get method 196ms 
        [+] returns false when the Test method is called 47ms 
        [+] throws exception as force isn't specified 47ms 
        [-] delete and create as force is specified 168ms 
          RuntimeException: You cannot call a method on a null-valued expression. 
          at <ScriptBlock>, C:\projects\xsharepoint\Modules\xSharePoint\DSCResources\MSFT_xSPUserProfileSyncConnection\MSFT_xSPUserProfileSyncConnection.psm1: line 154 
          at <ScriptBlock>, <No file>: line 2 
          at <ScriptBlock>, C:\Program Files\WindowsPowerShell\Modules\Pester\Functions\Mock.ps1: line 1056 
          at ExecuteBlock, C:\Program Files\WindowsPowerShell\Modules\Pester\Functions\Mock.ps1: line 1060 
          at Invoke-Mock, C:\Program Files\WindowsPowerShell\Modules\Pester\Functions\Mock.ps1: line 906 
          at <ScriptBlock><Process>, <No file>: line 70 
          at Set-TargetResource, C:\projects\xsharepoint\Modules\xSharePoint\DSCResources\MSFT_xSPUserProfileSyncConnection\MSFT_xSPUserProfileSyncConnection.psm1: line 85 
          at <ScriptBlock>, C:\projects\xsharepoint\Tests\xSharePoint\xSharePoint.xSPUserProfileSyncConnection.Tests.ps1: line 252 

This has been worked around by a change to force the build server to use Pester 3.3.14 instead of the latest version, but this should be investigated and resolved so we can go back to using the latest version of Pester for our tests.

@camiloborges I would love your input on this one since it was your resource originally. I'll investigate when I get the time but you might see this quicker.

**Comment by camiloborges** [comment]
Good timing. Took the afternoon off to be on my own in a pub. Now I have something else to do while having a sip or 20

Camilo Borges
Sent via iPhone while out and about

**Comment by camiloborges** [comment]
hey
Module: xSPUPSConn.
line 157:        $list = New-Object System.Collections.Generic.List[[Microsoft.Office.Server.UserProfiles.DirectoryServiceNamingContext]]

 
this should be returning the object defined on  test script, line 107-109. It is returning null, so $list doesn't have the 'Add' method


I am running out of time here, but got to it.

**Comment by andy-cox** [comment]
I've found I'm getting errors when using this resource if the SharePoint admin account is not a User Profile Service Application admin and if it does not have full control permissions to the Service App.

Is [this](http://sharepoint.stackexchange.com/questions/22161/powershell-error-when-trying-to-hook-into-userprofileconfigmanager) related to this problem?

The errors i'm getting:

SPAdmin not got Full Control permissions:
```powershell
PowerShell DSC resource MSFT_xSPUserProfileSyncConnection  failed to execute Test-TargetResource functionality with error message: Exception calling ".ctor" with "1" argument(s): "UserProfileApplicationNotAvailableException_Logging :: 
UserProfileApplicationProxy.ApplicationProperties ProfilePropertyCache does not have ad9....e56b661f" 
    + CategoryInfo          : InvalidOperation: (root/Microsoft/...gurationManager:String) [], CimException
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure
    + PSComputerName        : TheBrain
```

SPAdmin got permissions but not an admin of the UPSA:
```powershell
PowerShell DSC resource MSFT_xSPUserProfileSyncConnection  failed to execute Set-TargetResource functionality with error message: You cannot call a method on a null-valued expression. 
    + CategoryInfo          : InvalidOperation: (:) [], CimException
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure
    + PSComputerName        : TheBrain
```

@BrianFarnhill , @camiloborges  Let me know if this is not related and I'll create a new issue.

**Comment by camiloborges** [comment]
Hello.
Yes, you do have problems if the permissions aren't there.

I am not sure we kept it, but there was an exception being thrown when the upsmanager is null and the message actually say it is is possibly due to permissions.

@Brianfarnhill recently found an issue in multi server farms whereby it would only work if running with the farm account, which was, I think, included in the latest release

Camilo Borges
Sent via iPhone while out and about

**Comment by BrianFarnhill** [comment]
@andy-cox @camiloborges Yea I did two farms for a customer this week that both started the UPS fine using this as the resource:

            xSPUserProfileSyncService UserProfileSyncService
            {  
                UserProfileServiceAppName = "User Profile Service Application"
                Ensure                    = "Present"
                FarmAccount               = $FarmAccount
                PsDscRunAsCredential      = $SPSetupAccount
                DependsOn                 = "[xSPUserProfileServiceApp]UserProfileService"
            }

Thats based off the code that is in the 0.12 release branch (I think, I may have actually pulled the dev branch in which I think @camiloborges fixed something in, I need to double check). 

I actually think you issue is something different though @andy-cox - Did you configure CredSSP support on your VM at all (using xCredSSP or manually?). What I saw happen this week was that if you don't have this the user profile service actually provisions out as the setup account, not the farm account like it should - which means the user profile service database has the setup account as the schema owner, not the farm account - and that breaks the UPS's ability to provision. I removed my UPS, configured CredSSP with the resources below and then ran it again, and it all worked fine.

        xCredSSP CredSSPServer { Ensure = "Present"; Role = "Server"; } 
        xCredSSP CredSSPClient { Ensure = "Present"; Role = "Client"; DelegateComputers = "*.contoso.local" }

The wildcard there is because we were lazy in this environment and most people will wanna lock that down more, but I digress.

So if there is an issue that Andy has come up with here, it's that the user profile service should stop you from provisioning it if its not done as the farm account - which is a different issue to the tests failing here, so warrants a new issue. Do you guys agree?

**Comment by andy-cox** [comment]
@BrianFarnhill, I have configured CredSSP using xCredSSP and was also lazy using a wildcard domain in the delegated computers. The SP_UserProfiles DB's owner is also the farm account, SPFarm.

My config:
```powershell
xSPUserProfileServiceApp UserProfileServiceApp
{
    Name                 = $SharePoint.UserProfileService.Name
    ApplicationPool      = $ServiceAppPoolName
    MySiteHostLocation   = $SharePoint.UserProfileService.MySiteUrl
    ProfileDBName        = $SharePoint.UserProfileService.ProfileDB
    ProfileDBServer      = $SqlServer.ServiceAppDatabaseServer
    SocialDBName         = $SharePoint.UserProfileService.SocialDB
    SocialDBServer       = $SqlServer.ServiceAppDatabaseServer
    SyncDBName           = $SharePoint.UserProfileService.SyncDB
    SyncDBServer         = $SqlServer.ServiceAppDatabaseServer
    FarmAccount          = $FarmAccount
    PsDscRunAsCredential = $InstallAccount
    DependsOn            = @(
        '[xSPServiceAppPool]MainServiceAppPool',
        '[xSPManagedMetaDataServiceApp]ManagedMetadataServiceApp',
        '[xSPSearchServiceApp]SearchServiceApp'
    )
}

xSPUserProfileSyncConnection UserProfileSyncConnection
{
    ConnectionCredentials = $ServicePoolManagedAccount
    ConnectionType        = 'ActiveDirectory'
    Name                  = 'Company Directory Services'
    Forest                = $ActiveDirectory.DomainName
    IncludedOUs           = $SharePoint.UserProfileService.IncludedOUs
    UserProfileService    = $SharePoint.UserProfileService.Name
    PsDscRunAsCredential  = $InstallAccount
    DependsOn             = @(
        '[xSPUserProfileServiceApp]UserProfileServiceApp'
        '[xSPUserProfileSyncService]UserProfileSyncService'
    )
}
```

All the parameters are pretty standard.

**Comment by BrianFarnhill** [comment]
@andy-cox Open this as a new issue, its not related to the failing tests that opened this one so lets look at it as a separate issue. Thanks dude

**Comment by BrianFarnhill** [comment]
@camiloborges I think I stumbled across a fix for this one while looking in to some other issues. I'm taking this one and will open a PR when it's ready

**Comment by camiloborges** [comment]
looking forwarding to a solution to this :)
have you mocked the Mock method? :D

**Comment by BrianFarnhill** [comment]
@camiloborges No but the solution isn't far off that :P 

Actually one of the problems (and there were a few) were with how those [ANSI] objects were resolving, so mocked those up differently, then did some mocking around the methods that spit out the lists and stuff in there too.

**Comment by BrianFarnhill** [comment]
Pull Request #298 fixes this and will be merged to dev shortly

**Comment by BrianFarnhill** [comment]
Fixed in #298 - closing
